// pareto-runtime/dist/esc/generics/functions/createDictionaryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDictionaryBuilder = void 0;
function createDictionaryBuilder() {
    var imp = {};
    return {
        add: function (key, value) {
            imp[key] = value;
        },
        toDictionary: function () {
            return {
                forEach: function (callback) {
                    Object.keys(imp).sort().forEach(function (key) { return callback(imp[key], key); });
                },
                getLookup: function () {
                    return {
                        getUnsafe: function (key) {
                            var entry = imp[key];
                            if (entry === undefined) {
                                throw new Error("no such entry: ".concat(key, ", options: ").concat(Object.keys(imp).join(", ")));
                            }
                            return entry;
                        },
                        with: function (key, ifFound, ifNotFound) {
                            var entry = imp[key];
                            if (entry === undefined) {
                                return ifNotFound(Object.keys(imp).sort());
                            }
                            return ifFound(entry);
                        },
                    };
                },
                find: function (key, ifFound, ifNotFound) {
                    var keys = [];
                    var entry = null;
                    Object.keys(imp).forEach(function (k) {
                        keys.push(k);
                        if (k === key) {
                            entry = imp[k];
                        }
                    });
                    if (entry === null) {
                        return ifNotFound(keys);
                    }
                    else {
                        return ifFound(entry);
                    }
                },
            };
        },
    };
}
exports.createDictionaryBuilder = createDictionaryBuilder;
